### PR TITLE
mango feeds connector filter improvement

### PR DESCRIPTION
**Pull request by grooviegermanikus**
This PR addresses the issue https://github.com/blockworks-foundation/mango-feeds/issues/6:
* makes filter configuration for the source stream more explicit to fit the required use cases
* simplifies snapshot handling

### On Snapshot
* Snapshot method implemented a switch for gPA/gMA which IMO is not necessary.
* Snapshot methods' return types exposed Response types from network library - I've added structs instead.

### On Filtering
new style:
* filtering is either by account_ids or program_id (**xor**)
* program_id is only one item (not a list)

old style:
* a list of account_ids and a list program_ids

all existing use cases and all implementation only require the proposed _new style_

the proposed new config style uses Rust ADT (FilterByAccountIds+FilterByProgramId) to allow either/or style of the filtering

### CODE REVIEW

File: connector/src/websocket_source.rs [hunk 1]
```diff
@@ -17,30 +17,165 @@ use std::{
...
+                                    .map_err_anyhow()?)).await.expect("sending must succeed");
+                    },
+                    None => {
+                        // note: this might loop if the filter did not return anything
+                        warn!("account stream closed");
+                        // that is weired - assume that it is okey to cancel in all cases
+                        // if filter_config.program_ids.is_empty() {
+                        //     return Ok(());
+                        // }
+                        return Ok(());
```
Comment (grooviegermanikus) [inline connector/src/websocket_source.rs:158]: note: that code is a bit unclear to me; assuming now that it is okey to cancel in all cases

File: connector/examples/geyser_example_consumer.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        loop {
+            let next = slot_queue_receiver.recv().await.unwrap();
+            // println!("got slot: {:?}", next);
+        }
+    });
+
+    tokio::spawn(async move {
+        loop {
+            let next = account_write_queue_receiver.recv().await.unwrap();
+            println!("got account write: {:?}", next);
```
Comment (godmodegalactus) [inline connector/examples/geyser_example_consumer.rs:69]: Same here.

File: connector/examples/config-yellowstone-grpc-testing.json [hunk 3]
```diff
@@ -0,0 +1,52 @@
+{
+    "libpath": "/Users/stefan/mango/code/yellowstone-grpc/target/debug/libyellowstone_grpc_geyser.dylib",
```
Comment (ckamm) [inline connector/examples/config-yellowstone-grpc-testing.json]: leaks some private path

File: connector/examples/geyser_example_consumer.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    };
+
+    let (account_write_queue_sender, account_write_queue_receiver) =
+        async_channel::unbounded::<AccountWrite>();
+
+    let (slot_queue_sender, slot_queue_receiver) = async_channel::unbounded::<SlotUpdate>();
+
+    tokio::spawn(async move {
+        loop {
+            let next = slot_queue_receiver.recv().await.unwrap();
```
Comment (godmodegalactus) [inline connector/examples/geyser_example_consumer.rs:61]: Task should exit on error.

File: connector/examples/websocket_example_consumer.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    let filter_config = filter_config1;
+
+    let (account_write_queue_sender, account_write_queue_receiver) =
+        async_channel::unbounded::<AccountWrite>();
+
+    let (slot_queue_sender, slot_queue_receiver) = async_channel::unbounded::<SlotUpdate>();
+
+    tokio::spawn(async move {
+        loop {
+            let next = slot_queue_receiver.recv().await.unwrap();
```
Comment (godmodegalactus) [inline connector/examples/websocket_example_consumer.rs:67]: same here.

File: connector/examples/websocket_example_consumer.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            // println!("got slot: {:?}", next);
+        }
+    });
+
+    tokio::spawn(async move {
+        loop {
+            let next = account_write_queue_receiver.recv().await.unwrap();
```
Comment (godmodegalactus) [inline connector/examples/websocket_example_consumer.rs:74]: same here.

File: connector/src/chain_data.rs [hunk 7]
```diff
@@ -229,7 +237,7 @@ impl ChainData {
             .ok_or_else(|| anyhow::anyhow!("account {} has no live data", pubkey))
     }
 
-    pub fn iter_accounts<'a>(&'a self) -> impl Iterator<Item = (&'a Pubkey, &'a AccountData)> {
```
Comment (godmodegalactus) [inline connector/src/chain_data.rs]: why did you remove this? May be some other libraries are using these lifetimes.

File: connector/src/grpc_plugin_source.rs [hunk 8]
```diff
@@ -383,14 +394,11 @@ pub async fn process_events(
...
                 );
-                let result = out.await;
-                assert!(result.is_err());
-                if let Err(err) = result {
-                    warn!(
-                        "error during communication with the geyser plugin. retrying. {:?}",
-                        err
-                    );
-                }
+                let err = out.await.unwrap_err();
```
Comment (godmodegalactus) [inline connector/src/grpc_plugin_source.rs]: if `let Err(err) ` is more used than unwrap_err.

File: connector/src/snapshot.rs [hunk 9]
```diff
@@ -9,12 +8,25 @@ use solana_client::{
...
+}
+
+/// gMA snapshot struct
+pub struct SnapshotMultipleAccounts {
+    pub slot: Slot,
+    // (account pubkey, snapshot account)
+    pub accounts: Vec<(String, Option<UiAccount>)>,
+}
 
 pub async fn get_snapshot_gpa(
```
Comment (godmodegalactus) [inline connector/src/snapshot.rs:26]: I am not sure changing the signature of the problem will have issues other libraries depending on this code.

File: connector/src/snapshot.rs [hunk 10]
```diff
@@ -31,19 +43,29 @@ pub async fn get_snapshot_gpa(
...
+        }
+        OptionalContext::NoContext(_) => anyhow::bail!("bad snapshot format"),
+    }
 }
 
 pub async fn get_snapshot_gma(
     rpc_http_url: String,
     ids: Vec<String>,
-) -> anyhow::Result<solana_client::rpc_response::Response<Vec<Option<UiAccount>>>> {
+) -> anyhow::Result<SnapshotMultipleAccounts> {
```
Comment (godmodegalactus) [inline connector/src/snapshot.rs:68]: Same here may break dependencies.

File: service-mango-fills/src/main.rs [hunk 11]
```diff
@@ -58,6 +59,7 @@ pub struct Peer {
     pub head_updates: bool,
 }
 
+#[allow(clippy::too_many_arguments)]
```
Comment (godmodegalactus) [inline service-mango-fills/src/main.rs:61]: You did not add any arguments why did you add this allow clippy.
Could be another commit to fix clippy messages.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: connector/src/lib.rs [hunk 12]
```diff
@@ -100,10 +100,15 @@ pub struct SnapshotSourceConfig {
     pub rpc_http_url: String,
 }
 
+#[derive(Clone, Debug, Deserialize)]
+pub enum EntityFilter {
+    FilterByAccountIds(Vec<String>),
+    FilterByProgramId(String),
```
Comment (ckamm) [inline connector/src/lib.rs]: not a required change since it was like this before, but it'd make more sense to me if this used `Pubkey` over `String`.

File: connector/src/lib.rs [hunk 13]
```diff
@@ -100,10 +100,15 @@ pub struct SnapshotSourceConfig {
...
+}
+
 #[derive(Clone, Debug, Deserialize)]
 pub struct FilterConfig {
-    pub program_ids: Vec<String>,
-    pub account_ids: Vec<String>,
+    pub entity_filter: EntityFilter,
```
Comment (ckamm) [inline connector/src/lib.rs:125]: this struct is for future-proofing?

Comment (ckamm) [reply connector/src/lib.rs:125]: If we were to use this for mango-v4 (the liquidator would really like to use grpc streams!), we'd need to be able to subscribe to:
- all mango-v4 program accounts
- a list of oracle accounts
- all open book v1 accounts that match a bunch of binary data filters

The idea is that you'd run multiple `source` instances to cover the three streams?

I guess that makes sense, the only share data between them is the slots stream as far as I know.

There may be some added difficulty with waiting for the three inputs to come up and be ready. Maybe later there could be a wrapper for complex setups like that.

File: connector/src/snapshot.rs [hunk 14]
```diff
@@ -55,50 +77,22 @@ pub async fn get_snapshot_gma(
         min_context_slot: None,
     };
 
-    info!("requesting snapshot {:?}", ids);
-    let account_snapshot = rpc_client
+    info!("requesting gma snapshot {:?}", ids);
+    let account_snapshot_response = rpc_client
         .get_multiple_accounts(ids.clone(), Some(account_info_config))
```
Comment (ckamm) [inline connector/src/snapshot.rs:82]: future: gma typically has an upper limit, might need to split `ids` and issue multiple parallel requests

File: connector/examples/geyser_example_consumer.rs (hunk 4 above)
Comment (grooviegermanikus) [reply connector/examples/geyser_example_consumer.rs:61]: unwrap will panic and terminate the tokio thread - no?

File: connector/src/lib.rs (hunk 12 above)
Comment (grooviegermanikus) [reply connector/src/lib.rs]: will go for Pubkey -  thought the same

File: connector/examples/geyser_example_consumer.rs (hunk 4 above)
Comment (godmodegalactus) [reply connector/examples/geyser_example_consumer.rs:61]: true true

File: connector/src/lib.rs (hunk 13 above)
Comment (grooviegermanikus) [reply connector/src/lib.rs:125]: thx for sharing the use cases.
my changes are guided by the current filter pattern. it might be that websocket+grpc supports the complexity you mentioned in one filter - yet I doubt it.

so yes, we have to support multiple sources with different filters and merge the stream data somehow.

File: connector/examples/geyser_example_consumer.rs (hunk 4 above)
Comment (godmodegalactus) [reply connector/examples/geyser_example_consumer.rs:61]: Nah sorry, unwrap will panic and stop the whole process right.

File: connector/src/snapshot.rs (hunk 9 above)
Comment (grooviegermanikus) [reply connector/src/snapshot.rs:26]: good point indeed!
checked the potential consumer but found none. would risk breaking downstream - fixing that would be straight forward

File: connector/src/snapshot.rs (hunk 10 above)
Comment (grooviegermanikus) [reply connector/src/snapshot.rs:68]: would risk to break it. the signature is so bad

File: service-mango-fills/src/main.rs (hunk 11 above)
Comment (grooviegermanikus) [reply service-mango-fills/src/main.rs:61]: true, I will isolate clippy fixes to a new PR

Comment (grooviegermanikus) [reply service-mango-fills/src/main.rs:61]: https://github.com/blockworks-foundation/mango-feeds/pull/13

Comment (grooviegermanikus) [reply service-mango-fills/src/main.rs:61]: done

File: connector/src/lib.rs (hunk 12 above)
Comment (grooviegermanikus) [reply connector/src/lib.rs]: done

File: connector/src/snapshot.rs (hunk 14 above)
Comment (grooviegermanikus) [reply connector/src/snapshot.rs:82]: good point. yet the code in _rpc.rs_ does not impose a limit AFAIK

File: connector/src/lib.rs (hunk 13 above)
Comment (grooviegermanikus) [reply connector/src/lib.rs:125]: updated the initial ticket wrt "multiplexing":
1. websocket+geyser
2. multiple of same type
3. multiple subscriptions(filters)

File: connector/src/chain_data.rs (hunk 7 above)
Comment (grooviegermanikus) [reply connector/src/chain_data.rs]: strange, looks like a GitHub UI glitch. I did not remove it - just removed the lifetime annotations.

this is the actual patch:

```
-    pub fn iter_accounts<'a>(&'a self) -> impl Iterator<Item = (&'a Pubkey, &'a AccountData)> {
+    pub fn iter_accounts(&self) -> impl Iterator<Item = (&Pubkey, &AccountData)> {
```
